### PR TITLE
decred: Add input_derive_script.

### LINKED ISSUE #4161: dcr: Cannot spend stake outputs.
**Describe the bug**
Decred shares code with bitcoin and a change broke signing stake based inputs.

Bug introduced by c9e6ca30c925a213bc92c7b283875d166d3c4f75

**Firmware version and revision**
Trezor T v2.5.1 ~ current

**To Reproduce**
Buy a ticket with Trezor, then try to spend the funds after the ticket is voted or revoked.

**Expected behavior**
We can spend the outputs.

### PR DESCRIPTION
**Pull request by JoeGruffins**
Decred shares code with bitcoin and a recent change broke signing stake
    based inputs. Add decred specific script construction and fix tests.

This fixes a bug introduced by c9e6ca30c925a213bc92c7b283875d166d3c4f75

This is partly my fault as the test input was flawed and did not have the proper script opcode prefix.

Feel free to take these changes and adjust as you like. I do not need credit. Just hope to get the bug fixed asap.

closes #4161

### CODE REVIEW

**Comment by JoeGruffins** [comment]
@matejcik Hey buddy. Any chance you could have a look at this. Almost a year old. Thanks.

**Comment by teknico** [comment]
IIUC, this change is needed to stake Decred coins. This is the remaining blocker for me and other Decred holders. We could finally use our Model T’s for Decred, after all these years.

**Comment by Hannsek** [comment]
Hello,
not yet, sorry. We hopefully take a look until the EOY.

**Comment by peterzen** [comment]
@matejcik please merge this.

**Comment by matthawkins90** [comment]
It's wild that this hasn't been reviewed or merged yet.

**Comment by areyoufooled** [comment]
@matejcik please help with this PR, lots of users are blocked by this

**Comment by Hannsek** [comment]
Hi, this has currently low priority as we don't see demand for that feature. Once we have free hands we'll take a look.

**Comment by areyoufooled** [comment]
> Hi, this has currently low priority as we don't see demand for that feature. Once we have free hands we'll take a look.

Appreciate the reply. Our community has been asking for this for years (PR been here for two years!!) and I promise there are many out there waiting for this feature. If manpower is a concern, our DAO may be able to compensate any cost related.

**Comment by exitus1** [comment]
Please merge this! Our community wants staking via a hardware wallet.

**Comment by dev-warrior777** [comment]
> Hi, this has currently low priority as we don't see demand for that feature. Once we have free hands we'll take a look.

Thank you for replying. What can we do to help review process along .. we want to help assist review.

May we help, please? Thanks in advance.

**Comment by JoeGruffins** [comment]
> Hi, this has currently low priority as we don't see demand for that feature. Once we have free hands we'll take a look.

This is not a feature it is a bug fix.

File: core/src/apps/bitcoin/scripts_decred.py
Comment (matejcik) [inline core/src/apps/bitcoin/scripts_decred.py:22]: please revert

File: core/src/apps/bitcoin/scripts_decred.py
Comment (matejcik) [inline core/src/apps/bitcoin/scripts_decred.py:28]: please convert to micropython consts:
```
from micropython import const

_OP_SSTX = const(0xBA)
...
```
that way micropython bytecode compiler will do a search&replace instead of doing a lookup in the global namespace, and leading `_` stops the symbols from taking up space in the module dict

**Comment by JoeGruffins** [comment]
Just rebased.

**Comment by JoeGruffins** [comment]
It's been a while and I don't seem to be able to build or run tests, but here are the requested changes, untested https://github.com/trezor/trezor-firmware/compare/0cc6235ad296bcc22bfe27fafe54e3e20dadbb65..87117221ccbefdbf7e5db6092df0a5bbf1b305d7

File: core/src/apps/bitcoin/scripts_decred.py
Comment (JoeGruffins) [reply core/src/apps/bitcoin/scripts_decred.py]: reverted

File: core/src/apps/bitcoin/scripts_decred.py
Comment (JoeGruffins) [reply core/src/apps/bitcoin/scripts_decred.py]: reverted

File: core/src/apps/bitcoin/scripts_decred.py
Comment (JoeGruffins) [reply core/src/apps/bitcoin/scripts_decred.py]: I changed these as requested.

**Comment by matejcik** [comment]
thanks, CI is running now, we'll see what falls out

File: core/src/apps/bitcoin/scripts_decred.py
Comment (JoeGruffins) [reply core/src/apps/bitcoin/scripts_decred.py:19]: Added the import and was able to build today. Unsure if it is in the right place... https://github.com/trezor/trezor-firmware/compare/87117221ccbefdbf7e5db6092df0a5bbf1b305d7..4700af2cb1aed90ba71c37494c84c0de28e22c2e

File: core/src/apps/bitcoin/sign_tx/decred.py
Comment (matejcik) [inline core/src/apps/bitcoin/sign_tx/decred.py:441]: silence the type checker:
```suggestion
        assert txi.decred_tree is not None and txi.decred_staking_spend is not None
        return scripts_decred.output_derive_script(
```

File: core/src/apps/bitcoin/scripts_decred.py
Comment (matejcik) [inline core/src/apps/bitcoin/scripts_decred.py:8]: should be moved before `from typing`, as the very first line. like here: https://github.com/trezor/trezor-firmware/blob/matejcik/coins-t3b1/core/src/apps/binance/helpers.py#L1

**Comment by JoeGruffins** [comment]
Thank you. Moved `const` import to the very top and asserting `not None` https://github.com/trezor/trezor-firmware/compare/4700af2cb1aed90ba71c37494c84c0de28e22c2e..b6edbab5436643315472c63618b74cddfe7f7467

File: core/src/apps/bitcoin/scripts_decred.py
Comment (matejcik) [inline core/src/apps/bitcoin/scripts_decred.py:175]: isort wants an empty line here:
```suggestion
    from trezor.crypto import base58

    from apps.common import address_type
```

**Comment by matejcik** [comment]
tests are failing. it appears that the asserts that I suggested are incorrect:
```
class TestSignTxDecred
  test_one_one_fee ... failed
Traceback (most recent call last):
  File "/home/runner/work/trezor-firmware/trezor-firmware/core/tests/unittest.py", line 250, in run_class
  File "test_apps.bitcoin.signtx_decred.py", line 210, in test_one_one_fee
  File "../src/apps/bitcoin/sign_tx/bitcoin.py", line 63, in signer
  File "../src/apps/bitcoin/sign_tx/bitcoin.py", line 205, in step1_process_inputs
  File "../src/apps/bitcoin/sign_tx/decred.py", line 441, in input_derive_script
AssertionError: 
  test_purchase_ticket ... failed
Traceback (most recent call last):
  File "/home/runner/work/trezor-firmware/trezor-firmware/core/tests/unittest.py", line 250, in run_class
  File "test_apps.bitcoin.signtx_decred.py", line 397, in test_purchase_ticket
  File "../src/apps/bitcoin/sign_tx/bitcoin.py", line 63, in signer
  File "../src/apps/bitcoin/sign_tx/bitcoin.py", line 205, in step1_process_inputs
  File "../src/apps/bitcoin/sign_tx/decred.py", line 441, in input_derive_script
AssertionError: 
Ran 2 tests (2 failed)
```
perhaps instead of these asserts, the called function should allow optional parameters? i dunno :woman_shrugging:

**Comment by JoeGruffins** [comment]
I need how to figure out how to build the emulator again to run the tests...

**Comment by JoeGruffins** [comment]
Hey thanks, sorry this is taking your time. I am able to run tests now. I'll learn how to lint if this one also does not work... https://github.com/trezor/trezor-firmware/compare/b6edbab5436643315472c63618b74cddfe7f7467..de6491ce1ca05865c564d4e4ca62b7f293230655

**Comment by JoeGruffins** [comment]
Thats not a bool is it, wait a minute.

**Comment by matejcik** [comment]
if what you need here is just default values, you could add them to protobuf definitions, then the type of the field is no longer optional
(you might need to modify some sanitize functions in helpers.py)

**Comment by JoeGruffins** [comment]
No I don't guess a default value is correct for the stake type, I thought it was a bool for some reason.

What check am I failing exactly? Would be good to run locally.

Anyway trying again https://github.com/trezor/trezor-firmware/compare/de6491ce1ca05865c564d4e4ca62b7f293230655..b843127b82fbed3df8cfb97fbdf6bcfdf6c2d68d

**Comment by matejcik** [comment]
try `make pyright` in `core/` or `make style` in root

**Comment by JoeGruffins** [comment]
`make style` changed the layout of arguments but `make pyright` shows a lot of stuff that does not seem relevant to my changes. https://github.com/trezor/trezor-firmware/compare/b843127b82fbed3df8cfb97fbdf6bcfdf6c2d68d..4655c028f3c57413f585d5a7205531b00fcf2e8e

**Comment by matejcik** [comment]
> but `make pyright` shows a lot of stuff that does not seem relevant to my changes

you're probably using pyright from your own environment, of some random version. pyright likes to introduce a lot of changes that our codebase triggers.
(honestly I might want to look into adding trezor-firmware to pyright's test corpus)

if you installed pyright via pip, `export PYRIGHT_PYTHON_FORCE_VERSION=1.1.361` will get you to where we're at

**Comment by JoeGruffins** [comment]
I should add an entry to `core/CHANGELOG.md` ?

**Comment by matejcik** [comment]
we do changelogs via towncrier: https://docs.trezor.io/trezor-firmware/misc/changelog.html
please prefix the change entry with `[T2T1]` because it only affects Trezor T

> I'm guessing the current three github test failures are not related to my changes?

right, we can ignore those

**Comment by JoeGruffins** [comment]
There wasn't an issue... should I make one for the change log?

**Comment by JoeGruffins** [comment]
Attempted to add the change log https://github.com/trezor/trezor-firmware/compare/cb9864586713acada2a147a6bb796522cb01ad03..7346063f75e22a80b66c49682f4e3f25e539005f

**Comment by matejcik** [comment]
> There wasn't an issue... should I make one for the change log?

you could have used the id of this PR. but this works too

**Review by matejcik — APPROVED** [review]

**Comment by matejcik** [comment]
and after almost two years, we're ready to merge. thanks for your patience!

**Comment by peterzen** [comment]
> and after almost two years, we're ready to merge. thanks for your patience!

Thank you!

**Pull request merged**
